### PR TITLE
Replace the deprecated event dispatching via RCTEventDispatcher with …

### PR DESCRIPTION
**Pull request by mars-lan**
#63 …RCTBubblingEventBlocks for iOS.

### CODE REVIEW

**Comment by mars-lan** [comment]
Should also fix #67

**Comment by sbugert** [comment]
Sorry for not merging your PR yet! I will do that tomorrow as soon as possible.

What is the min required react-native version for this? Could you also change the value in the readme to reflect that? :)

**Comment by mars-lan** [comment]
The relevant API was introduced prior to 0.19 so there should be no need to update readme.

**Comment by emrecanozkok** [comment]
is this #67  allready fixed ?

**Comment by avishayil** [comment]
Does not solving the #67 problem for me.
In fact, it's only causing the app to crash at load.

**Comment by Jeiwan** [comment]
This PR fixes https://github.com/sbugert/react-native-admob/issues/67 for me, as well as https://github.com/sbugert/react-native-admob/issues/63.

But the problem is that when `adViewDidReceiveAd` property is set, the ad is not shown. The event is fired successfully, but the ad is not rendered and the view that holds the ad is not rendered as well. Interesting thing is that commenting out `self.onAdViewDidReceiveAd(@{});` in `- (void)adViewDidReceiveAd:(DFPBannerView *)adView` doesn't solve the issue, only removing `adViewDidReceiveAd` property from `PublisherBanner` component makes it work again.

**Comment by mars-lan** [comment]
@avishayil Do you have the stack trace?
@Jeiwan I'm unable to reproduce the issue from my end. The ad shows just fine with `adViewDidReceiveAd` property set.

**Comment by Jeiwan** [comment]
@mars-lan I found out that the problem occurs only when `testDeviceID` is set to some value (not `undefined`). This happens when I set it to `EMULATOR` or `null`. Could you try setting it please?

**Comment by mars-lan** [comment]
@Jeiwan Hum, strange. I was able to show the test ads by adding `testDeviceID="EMULATOR"` without any issue on RN 0.38 & 0.39. Could you try a really simple example to rule out other possibilities?

**Comment by Jeiwan** [comment]
@mars-lan I created a dummy app that reproduces the bug: https://github.com/Jeiwan/wootadmob It uses cocoapods so you need to open `wootadmob.xcworkspace` in Xcode.
There's a red label that is changed to "AD IS LOADED" when `adViewDidReceiveAd` is triggered. If `testDeviceID` is set to any value, the label is updated, but the ad is not shown. If `testDeviceID` is not set, the label is updated and the ad is shown. If `adViewDidReceiveAd` is not set, the label is not updated and the ad is shown even if `testDeviceID` is set. So it looks like there's a conflict when both `testDeviceID` and `adViewDidReceiveAd` are set.

**Comment by mars-lan** [comment]
Argh I see. I've been using `AdMobBanner` instead of `PublisherBanner`. Can finally reproduce it now.

**Comment by avishayil** [comment]
App is crashing using this PR:
````
dyld: Symbol not found: ___NSDictionary0__
  Referenced from: /Users/avishay/Library/Developer/CoreSimulator/Devices/85498380-AE5B-4F30-962C-F8F3B813C154/data/Containers/Bundle/Application/E998399F-AC47-49FA-8E30-6AEECF54731C/Geektime.app/Geektime
  Expected in: /Library/Developer/CoreSimulator/Profiles/Runtimes/iOS 8.1.simruntime/Contents/Resources/RuntimeRoot/System/Library/Frameworks/CoreFoundation.framework/CoreFoundation
 in /Users/avishay/Library/Developer/CoreSimulator/Devices/85498380-AE5B-4F30-962C-F8F3B813C154/data/Containers/Bundle/Application/E998399F-AC47-49FA-8E30-6AEECF54731C/Geektime.app/Geektime
````

**Comment by mars-lan** [comment]
@avishayil Not sure if it's related to the PR or your dev environment: http://stackoverflow.com/questions/32815121/dyld-symbol-not-found-nsdictionary0-when-using-google-servicegenerator-bi

Are you using the latest XCode + macOS?

**Comment by avishayil** [comment]
Yes I do use the latest setup.
Regular library without this PR works perfectly.
Do you have any idea how to find out which line of code causes the crash? I think I'll try replacing line by line and try to find it out.
By the way - I'm using PublisherBanner on my app.

**Comment by atominvention** [comment]
@avishayil I have the same crash after using this PR
Then I open Xcode, under "Build Phases" => "Link Binary With Libraries", add "CoreFoundation" as optional, then it works

**Comment by alvaromb** [comment]
Hi @mars-lan!

How do you feel about updating your PR?

**Comment by mars-lan** [comment]
@alvaromb This should be ready to merge again.

**Comment by alvaromb** [comment]
Will review as soon as I can, thanks @mars-lan !

**Comment by Sh3rawi** [comment]
The problem of the banner ads not rendering in iOS faced me when i upgraded from RN 0.37 to 0.38, i installed this fork to see if it would fix it, but iOS did not build successfully with me until i did the following fixes:
1)  BannerView.h:
```
#if __has_include(<React/RCTComponent.h>)
#import <React/RCTComponent.h>
#else
#import "RCTComponent.h"
#endif
@import GoogleMobileAds;
```

instead of 
```
#if __has_include(<React/RCTEventDispatcher.h>)
#import <React/RCTEventDispatcher.h>
#else
#import "RCTEventDispatcher.h"
#endif
@import GoogleMobileAds;
@class RCTEventDispatcher;
```

2) RNDFPBannerView.h:
```
#if __has_include(<React/RCTEventDispatcher.h>)
#import <React/RCTComponent.h>
#else
#import "RCTComponent.h"
#endif
@import GoogleMobileAds;
@class RCTEventDispatcher;
@interface RNDFPBannerView : UIView <GADBannerViewDelegate>
```

instead of:
```
#if __has_include(<React/RCTEventDispatcher.h>)
#import <React/RCTEventDispatcher.h>
#else
#import "RCTEventDispatcher.h"
#endif
@import GoogleMobileAds;
@class RCTEventDispatcher;
@interface RNDFPBannerView : UIView <GADBannerViewDelegate>
```

After these edits, i was able to successfully build it, and the banner finally rendered. However, i do not know if anything broke with the DFPBanner because i only use regular banners.

**Comment by mars-lan** [comment]
Thanks @Sh3rawi. The import was broken due to a bad merge. Should be building fine on RN 0.38-0.42 now.

**Comment by Sh3rawi** [comment]
@mars-lan you're welcome. All is well now 👍

**Comment by mars-lan** [comment]
@alvaromb Please merge before this diverges further again thanks!

**Comment by alvaromb** [comment]
Reviewing right now. Thanks @mars-lan!

**Comment by mars-lan** [comment]
@alvaromb Done. PTAL.

File: ios/RNAdMobInterstitial.m [hunk 1]
```diff
@@ -71,27 +69,37 @@ - (dispatch_queue_t)methodQueue
 #pragma mark delegate events
 
 - (void)interstitialDidReceiveAd:(GADInterstitial *)ad {
-  [self.bridge.eventDispatcher sendDeviceEventWithName:@"interstitialDidLoad" body:nil];
+  if (self.onInterstitialDidLoad) {
```
Comment (alvaromb) [inline ios/RNAdMobInterstitial.m]: Hi @mars-lan!

Sorry for not being sufficiently clear, this doesn't work if the handler is not part of a `View`. This is why it works with the Banner.

**Comment by alvaromb** [comment]
Some RN features are not clearly documented, as you can see. ¯\_(ツ)_/¯

**Comment by mars-lan** [comment]
@alvaromb Thanks for the info. Unfortunately this seems like a much larger change now, especially given that there doesn't seem to be a corresponding mechanism for Android. I've reverted my last commit.

How about we get this merged in first and tackle Interstitial & Rewarded video in a separate PR?

**Comment by alvaromb** [comment]
I'm cool with that, I'll merge once I get to the office. Thanks!

Enviado desde mi iPhone

**Comment by mars-lan** [comment]
Ping?

**Comment by alvaromb** [comment]
Merged! Thanks @mars-lan.

**Pull request merged**